### PR TITLE
【Hackathon 5th No.26】为 Paddle 新增 diagonal_scatter API -part

### PR DESCRIPTION
**Pull request by DanGuge**
### PR types

Others
### PR changes

Others
### Description

为 Paddle 新增 diagonal_scatter API
RFC：[PaddlePaddle/community#669](https://github.com/PaddlePaddle/community/pull/669)
#57262

### CODE REVIEW

**Comment by DanGuge** [comment]
@zxcd 辛苦review一下，diagonal_scatter的实现

File: python/paddle/tensor/manipulation.py [hunk 1]
```diff
@@ -5575,3 +5575,140 @@ def index_fill_(x, index, axis, value, name=None):
...
+            'paddle.tensor.manipulation.diagonal_scatter',
+        )
+        check_variable_and_dtype(
+            y,
+            'Y',
+            [
+                'float16',
+                'float32',
+                'float64',
+                'bfloat16',
```
Comment (zxcd) [inline python/paddle/tensor/manipulation.py]: bfloat16目前在这里应该是使用uint16?

File: python/paddle/tensor/manipulation.py [hunk 2]
```diff
@@ -5575,3 +5575,140 @@ def index_fill_(x, index, axis, value, name=None):
...
+    assert tuple(predshape) == tuple(
+        y.shape
+    ), f"y.shape should be {tuple(predshape)}, but received {tuple(y.shape)}"
+    if len(y.shape) == 1:
+        y = y.reshape([1, -1])
+
+    if in_dynamic_mode():
+        return _C_ops.fill_diagonal_tensor(x, y, offset, axis1, axis2)
+    else:
+        helper = LayerHelper('diagonal_scatter', **locals())
```
Comment (zxcd) [inline python/paddle/tensor/manipulation.py]: 建议放在check_variable_and_dtype后面

File: test/legacy_test/test_diagonal_scatter.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        self.x = np.random.random([10, 10]).astype(np.float32)
+        self.y = np.random.random([10]).astype(np.float32)
+        self.offset = 0
+        self.axis1 = 1
+        self.axis2 = 0
+
+
+# check error
+class TestDiagonalScatterError(TestDiagonalScatterAPI):
+    def test_error_1(self):
```
Comment (zxcd) [inline test/legacy_test/test_diagonal_scatter.py]: 改成有意义的命名

**Comment by DanGuge** [comment]
@zxcd 修改完了，辛苦再review一下

**Review by zxcd — APPROVED** [review]
LGTM

File: python/paddle/__init__.py [hunk 4]
```diff
@@ -256,6 +256,7 @@
     masked_fill_,
     index_fill,
     index_fill_,
+    diagonal_scatter,
```
Comment (jeff41404) [inline python/paddle/__init__.py:263]: need to be added in list of \_\_all\_\_ below

**Comment by jeff41404** [comment]
The logic of the `diagonal_scatter` API and the existing function `fill_diagonal_tensor` in `python/paddle/tensor/manipulation.py` are most of the same, whether it can be directly exposed as an alias of `fill_diagonal_tensor` function, so no need to add additional Python logic? Meanwhile, its underlying operator `fill_ Diagonal_ tensor` needs to add support for `int16` data types

**Comment by DanGuge** [comment]
> The logic of the `diagonal_scatter` API and the existing function `fill_diagonal_tensor` in `python/paddle/tensor/manipulation.py` are most of the same, whether it can be directly exposed as an alias of `fill_diagonal_tensor` function, so no need to add additional Python logic? Meanwhile, its underlying operator `fill_ Diagonal_ tensor` needs to add support for `int16` data types

@jeff41404 
1. There is still difference between `fill_diagonal_tensor` and `diagonal_scatter`, the former one is only available in **dynamic mode** while the latter one supports both **dynamic & static mode**. And I think maybe we should not change the origin usage of `fill_diagonal_tensor`, so it may be better to add additional API for `diagonal_scatter`.
2. I can submit another PR for `int16` data types of `fill_diagonal_tensor` & add relative changes, e.g. tests and so on.

**Comment by jeff41404** [comment]
> 1. There is still difference between `fill_diagonal_tensor` and `diagonal_scatter`, the former one is only available in **dynamic mode** while the latter one supports both **dynamic & static mode**. And I think maybe we should not change the origin usage of `fill_diagonal_tensor`, so it may be better to add additional API for `diagonal_scatter`.

Adding branch of static mode to the current API `fill_diagonal_tensor` is the compatibility upgrade and correct approach, which can reduce future maintenance costs, as long as it is clearly described in the rfc document and pass existing test case.

**Comment by jeff41404** [comment]
> 2\. I can submit another PR for `int16` data types of `fill_diagonal_tensor` & add relative changes, e.g. tests and so on.

OK, you can add link of PR to the description above and add these tasks to rfc

**Comment by DanGuge** [comment]
@jeff41404 cc @zxcd 
I added the logic of static mode to `fill_diagonal_tensor` and simplified `diagonal_scatter`.
And I added `int16` to `fill_diagonal_tensor` in this PR, please review it.

**Review by jeff41404 — DISMISSED** [review]
LGTM

File: python/paddle/tensor/manipulation.py [hunk 5]
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
...
+
+def diagonal_scatter(x, y, offset=0, axis1=0, axis2=1, name=None):
+    """
+    Embed the values of Tensor ``y`` into Tensor ``x`` along the diagonal elements
+    of Tensor ``x``, with respect to ``axis1`` and ``axis2``.
+
+    This function returns a tensor with fresh storage.
+
+    The argument ``offset`` controls which diagonal to consider:
+    If ``offset`` = 0, it is the main diagonal.
```
Comment (ooooo-create) [inline python/paddle/tensor/manipulation.py]: 加上 - 更直观，下同

File: python/paddle/tensor/manipulation.py [hunk 6]
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
...
+    If ``offset`` > 0, it is above the main diagonal.
+    If ``offset`` < 0, it is below the main diagonal.
+
+    Note:
+        ``y`` should have the same shape as paddle.diagonal(x, offset, axis1, axis2).
+
+    Args:
+        x (Tensor): `x`` is the original Tensor. Must be at least 2-dimensional.
```
Comment (ooooo-create) [inline python/paddle/tensor/manipulation.py]:
```suggestion
        x (Tensor): ``x`` is the original Tensor. Must be at least 2-dimensional.
```

File: python/paddle/tensor/manipulation.py [hunk 7]
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
...
+        y (Tensor): ``y`` is the Tensor to embed into ``x``
+        offset (int, optional): which diagonal to consider. Default: 0 (main diagonal).
+        axis1 (int, optional): first axis with respect to which to take diagonal. Default: 0.
+        axis2 (int, optional): second axis with respect to which to take diagonal. Default: 1.
+        name (str, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Tensor, Tensor with diagonal embedeed with y.
```
Comment (ooooo-create) [inline python/paddle/tensor/manipulation.py]:
```suggestion
        Tensor, Tensor with diagonal embedeed with ``y``.
```

File: python/paddle/tensor/manipulation.py [hunk 8]
```diff
@@ -5575,3 +5632,46 @@ def index_fill_(x, index, axis, value, name=None):
...
+
+    Examples:
+        .. code-block:: python
+
+            >>> # doctest: +REQUIRES(env:GPU)
```
Comment (ooooo-create) [inline python/paddle/tensor/manipulation.py]: 这个是一定需要 GPU吗，如果不是，改成使用CPU的示例代码比较好~

File: python/paddle/tensor/manipulation.py [hunk 9]
```diff
@@ -5770,3 +5827,45 @@ def index_fill_(x, index, axis, value, name=None):
...
+
+
+def diagonal_scatter(x, y, offset=0, axis1=0, axis2=1, name=None):
+    """
+    Embed the values of Tensor ``y`` into Tensor ``x`` along the diagonal elements
+    of Tensor ``x``, with respect to ``axis1`` and ``axis2``.
+
+    This function returns a tensor with fresh storage.
+
+    The argument ``offset`` controls which diagonal to consider:
```
Comment (ooooo-create) [inline python/paddle/tensor/manipulation.py:5839]: 下面加个空行~

File: python/paddle/tensor/manipulation.py [hunk 10]
```diff
@@ -5770,3 +5827,45 @@ def index_fill_(x, index, axis, value, name=None):
...
+    - If ``offset`` = 0, it is the main diagonal.
+    - If ``offset`` > 0, it is above the main diagonal.
+    - If ``offset`` < 0, it is below the main diagonal.
+
+    Note:
+        ``y`` should have the same shape as paddle.diagonal(x, offset, axis1, axis2).
```
Comment (ooooo-create) [inline python/paddle/tensor/manipulation.py]: 写成这样会不会更好~
```suggestion
        ``y`` should have the same shape as :ref:`paddle.diagonal <api_paddle_diagonal>`.
```

**Review by ooooo-create — COMMENTED** [review]
其余LGTM @sunzhongkai588

**Review by ooooo-create — APPROVED** [review]
LGTM @sunzhongkai588

**Review by sunzhongkai588 — APPROVED** [review]
LGTM for docs

**Comment by luotao1** [comment]
hi, @DanGuge 
- 非常感谢你对飞桨框架的贡献，我们正在运营一个PFCC组织，会通过定期分享技术知识与发布开发者主导任务的形式持续为飞桨框架做贡献，详情可见 https://github.com/luotao1 主页说明。
- 如果你对PFCC有兴趣，请发送邮件至 [ext_paddle_oss@baidu.com](mailto:ext_paddle_oss@baidu.com)，我们会邀请你加入~

**Pull request merged**